### ISSUE TITLE
Set preferences seems not working

### ISSUE THREAD

**Issue by disassembly** [rmartin16/qbittorrent-api#8]
qbittorrent is 4.1.5, api version is 2.2, with python2.7
I follow the instructions:
">>> from qbittorrentapi import Client
">>> client = Client(host='localhost:8080', username='admin', password='adminadmin')
">>> is_dht_enabled = client.application.preferences.dht
">>>is_dht_enabled
True
">>> client.application.preferences = dict(dht=(not is_dht_enabled))
">>> client.application.preferences.dht
True
The results show that the set preferences is not working.
So I tried the set_preferences methods:
">>> client.app.set_preferences(dict(dht=(not is_dht_enabled)))
<Response [200]>
">>> client.application.preferences.dht
True
It is also not working.
Am I wrong to use the methods?
Best wishes!

**Comment by disassembly** [comment]
However, it works on qbittorrent 4.2.2 with an api 2.4.1.
Confusing.........

**Comment by rmartin16** [comment]
Thanks for reporting this. This one was a little annoying once I figured it out.

tl;dr: Older versions of qBittorrent's Web API apparently did not support whitespace in the JSON payload of the POST request for updating preferences. Removing the whitespace allows this client to update preferences in older qBittorrent versions.

After replicating it, it definitely was confusing. The issue extends beyond updating DHT support; it is not possible to update any preferences in qBittorrent v4.1.5. So, I spun up Wireshark and captured the packets from both qBittorrent's WebUI and this client for `setPreferences`. The WebUI request included the `X-Requested-With` header but that didn't make a difference when I added it to this client's request. The only other difference was the WebUI request did not include any whitespace in the JSON representation of the preferences. qBitorrent requires the JSON be set in a slightly peculiar fashion since they aren't looking for just a JSON payload in the POST request....they want a Form submission with the JSON payload set as the value for a 'json' field. At any rate, once I updated the JSON generation to use separators without any whitespace, qBittorrent started updated the preferences. I'm guessing qBittorrent started using a more robust JSON interpreter since v4.1.5.

commit forthcoming...

**Comment by disassembly** [comment]
Thanks very much for your effort of digging the problem. The Debian 9 or
Armbian 20.02.1 Buster (based on Debian 10) are using qbittorrent v4.1.5 in
their official repository. I have moved to qbittorrent v4.1.9 after a
compilation, and it works so cool with extra sync.maindata() support. Maybe
the Debian official repository will update the package sometime since the
newest version is already v4.2.3.
Qbittorrent-api makes it really easy to manage qbittorrent, well done!
Best wishes!

rmartin16 <notifications@github.com> 于2020年4月20日周一 上午5:11写道：
